src/storage/mutate/AddEdgesProcessor.cpp

```diff
@@ -13,27 +13,163 @@ namespace nebula {
+                data.emplace_back(std::move(key), std::move(edge.get_props()));
+            });
+            doPut(spaceId_, partId, std::move(data));
         });
-        doPut(spaceId, partId, std::move(data));
+    } else {
+        std::for_each(req.parts.begin(), req.parts.end(), [&](auto& partEdges){
+            auto partId = partEdges.first;
+            const auto &edges = partEdges.second;
+            auto atomic = [&]() -> std::string {
+                return addEdges(version, partId, edges);
+            };
+            auto callback = [partId, this](kvstore::ResultCode code) {
+                handleAsync(spaceId_, partId, code);
+            };
+            this->kvstore_->asyncAtomicOp(spaceId_, partId, atomic, callback);
+        });
+    }
+}
+
+std::string AddEdgesProcessor::addEdges(int64_t version, PartitionID partId,
+                                        const std::vector<cpp2::Edge>& edges) {
+    std::unique_ptr<kvstore::BatchHolder> batchHolder = std::make_unique<kvstore::BatchHolder>();
+
+    /*
+     * Define the map newIndexes to avoid inserting duplicate edge.
+     * This map means :
+     * map<edge_unique_key, pair<edge_prop, IndexItem>> ,
+     * -- edge_unique_key is only used as the unique key , for example:
+     * insert below edges in the same request:
+     *     kv(part1_src1_edgeType1_rank1_dst1 , v1)
+     *     kv(part1_src1_edgeType1_rank1_dst1 , v2)
+     *     kv(part1_src1_edgeType1_rank1_dst1 , v3)
+     *     kv(part1_src1_edgeType1_rank1_dst1 , v4)
+     *
+     * Ultimately, kv(part1_src1_edgeType1_rank1_dst1 , v4) . It's just what I need.
+     */
+    std::map<std::string, std::pair<std::string, nebula::cpp2::IndexItem>> newIndexes;
+    std::for_each(edges.begin(), edges.end(), [&](auto& edge){
+        auto prop = edge.get_props();
+        auto type = edge.key.edge_type;
+        auto srcId = edge.key.src;
+        auto rank = edge.key.ranking;
+        auto dstId = edge.key.dst;
+        VLOG(3) << "PartitionID: " << partId << ", VertexID: " << srcId
+                << ", EdgeType: " << type << ", EdgeRanking: " << rank
+                << ", VertexID: " << dstId << ", EdgeVersion: " << version;
+        for (auto& index : indexes_) {
+            /**
+             * skip the in-edge.
+             **/
+            if (index.get_tagOrEdge() == type) {
+                auto ekVal = NebulaKeyUtils::edgeKey(partId, srcId, type, rank, dstId, version);
+                newIndexes[ekVal] = std::make_pair(edge.get_props(), index);
+            }
+        }
     });
+    for (auto& index : newIndexes) {
+        /*
+         * step 1 , Delete old version index if exists.
+         */
+        auto oi = obsoleteIndex(partId, index.first, index.second.second);
+        if (!oi.empty()) {
+            batchHolder->remove(std::move(oi));
+        }
+        /*
+         * step 2 , Insert new edge data
+         */
+        auto key = index.first;
```

Comment:
auto&

Reply:
> auto&

Aha, because the argument of method `batchHolder->put` is an rvalue reference ，Here I use std::move to cast the reference.  So `auto&` doesn't work.